heron/instance/src/java/com/twitter/heron/instance/AbstractOutputCollector.java

```diff
@@ -93,23 +94,30 @@ public long getTotalTuplesEmitted() {
   }
 
   protected HeronTuples.HeronDataTuple.Builder initTupleBuilder(String streamId,
-                                                                List<Object> tuple) {
+                                                                List<Object> tuple,
+                                                                Integer taskId) {
     // Start construct the data tuple
     HeronTuples.HeronDataTuple.Builder builder = HeronTuples.HeronDataTuple.newBuilder();
 
     // set the key. This is mostly ignored
     builder.setKey(0);
 
     List<Integer> customGroupingTargetTaskIds = null;
-    if (!helper.isCustomGroupingEmpty()) {
+    if (taskId != null) {
+      // TODO: somehow assert that the input stream of the downstream bolt was configured
+      // with directGrouping
+
+      customGroupingTargetTaskIds = new ArrayList<>();
```

Comment:
@nlu90 I tried splitting up the code paths and it got pretty redundant and hard to follow. 

To remove the if/then I'd need to do the emit direct check at a higher level when creating the tuple builder. But the tuple builder is created in both SpoutOutputCollectorImpl.admitSpoutTuple and BoltOutputCollectorImpl.admitBoltTuple after doing other validation. To do the emit direct check separately, I'd need to have 4 more implementations of the same validation (with/without emit direct ids for Bolt/Spout collectors), and then support two similar code paths in AbstractOutputCollector to support emit direct, and non-emit direct. The latter impl would also have the custom grouping checks at 2 levels lower than the emit direct checks.

I can show you what I mean if it's not clear, but if you have a specific suggestion for how to avoid what I'm describing please let me know.